src/modules/tools/zprobe/CalibrationStrategy.cpp

```diff
@@ -0,0 +1,651 @@
+                } else {
+                    p++;
+                }
+            }
+
+            update_compensation_transformation(); // make sure no compensation transform from somewhere else is installed
+
+            gcode->stream->printf("Commencing calibration of parameters: %s\n", parameters_to_optimize.c_str());
+
+            samples = std::max(samples, (int)parameters_to_optimize.length());
+
+            if (optimize_model(samples, repeats, parameters_to_optimize, gcode->stream)) {
+                gcode->stream->printf("Calibration complete. Save settings with M500.\n");
+            } else {
+                gcode->stream->printf("Calibration may not have converged. Use M500 if you want to save settings anyway.\n");
+            }
+            THEKERNEL->call_event(ON_IDLE);
+
+            // Check if endstop trim was updated
+            if (parameters_to_optimize.find('X') != std::string::npos ||
+                parameters_to_optimize.find('Y') != std::string::npos ||
+                parameters_to_optimize.find('Z') != std::string::npos) {
+
+                // Deficiency: homing is the only way to activate the endstop trim values
+                zprobe->home();
+            }
+
+            return true;
+        }
+    } else if(gcode->has_m) {
+        // handle mcodes
+        if((gcode->m == 500 && (plane_u != 0 || plane_v != 0 || plane_offset != 0))
+                || gcode->m == 503) { // M500 save, M503 display
+            gcode->stream->printf(";Plane tilt:\n");
+            gcode->stream->printf("M567 U%.5f V%.5f W%.5f\n", plane_u, plane_v, plane_offset);
+            return true;
+        } else if(gcode->m == 567) {
+            if (gcode->has_letter('U')) plane_u = gcode->get_value('U');
+            if (gcode->has_letter('V')) plane_v = gcode->get_value('V');
+            if (gcode->has_letter('W')) plane_offset = gcode->get_value('W');
+
+            update_compensation_transformation();
+
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool CalibrationStrategy::handleConfig()
+{
+    // default is probably wrong
+    float r= THEKERNEL->config->value(leveling_strategy_checksum, calibration_strategy_checksum, radius_checksum)->by_default(-1.0f)->as_number();
+    if(r == -1) {
+        r= THEKERNEL->config->value(leveling_strategy_checksum, delta_calibration_strategy_checksum, radius_checksum)->by_default(65.0f)->as_number();
+    }
+    this->probe_radius= r;
+    return true;
+}
+
+
+bool CalibrationStrategy::setup_probe() {
+    // home
+    zprobe->home();
+
+    int s;
+    if(!zprobe->run_probe(s, false)) return false;
+
+    // We can not do a relative move now since we have lost our bearings (robot->last_milestone is incorrect)
+    // We also can not do a G92 since that resets our actuator positions.
+    // Luckily, our actuators know where they are.
+    THEKERNEL->robot->reset_position_from_current_actuator_position();
```

Comment:
run_probe (without returning or homing) requires this to allow movement after a probe result without resetting the actuator positions. I assume last_milestone is not being updated during run_probe().